blog/2024-01-24-transforming-your-data-1/index.md

```diff
@@ -0,0 +1,472 @@
+
+The data for the edges require minor transformations to the existing CSV files in which the first
+and second columns respectively represent the `FROM` and `TO` nodes' primary keys. To help reduce
+the amount of custom code you have to write, Kùzu provides convenient APIs to
+scan/read from CSV files and to copy data from CSV files to a node or edge table. An example is
+shown below.
+
+<div class="img-center">
+<img src={EdgeTables}/>
+</div>
+
+The `transaction.csv` file contains the `client_id` and `merchant_id` columns, which are the `FROM`
+and `TO` nodes' primary keys respectively, but these are not present in the first and second columns
+as the edge table requires. To do this, `transaction.csv` file isn't loaded as-is, but is instead
+transformed into the `transacted_with.csv` file via the following query.
+
+```sql
+-- Generate `transacted_with.csv` from `transaction.csv`
+COPY (
+    LOAD FROM 'transaction.csv' (header=true)
+    RETURN
+        client_id,
+        merchant_id,
+        amount_usd,
+        timestamp
+)
+TO 'transacted_with.csv';
+```
+
+The example above consists of two subquery blocks: `LOAD` and `COPY`. The `LOAD` block is used to
+scan the CSV file, check for headers and data types and return the columns we need. The `COPY`
+block is used to write the results we obtained from the scan to a new file.
+
+We can do the same for the other two edge tables as well.
+
+```sql
+// Generate `belongs_to.csv` from `merchant.csv`
+COPY (
+    LOAD FROM 'merchant.csv' (header=true)
+    RETURN
+        merchant_id,
+        company_id
+)
+TO 'belongs_to.csv';
+```
+
+```sql
+// Generate `located_in.csv` from `merchant.csv`
+COPY (
+    LOAD FROM 'merchant.csv' (header=true)
+    RETURN
+        merchant_id,
+        city_id
+)
+TO 'located_in.csv';
+```
+
+With all the input files in place, we can now proceed to insert the data and build the graph!
+
+## Insert data into Kùzu
+
+Collecting all the above functions, we can write a script that performs the following:
+
+1. Transform node/edge tables as needed
+1. Create the node tables
+1. Create edge tables
+1. Load the node tables into the database
+1. Load the edge tables into the database
+
+In Python, it would look something like this:
+
+```sql
+// Copy from CSV to node tables
+COPY Client FROM 'client.csv' (header=true)
+COPY City FROM 'city.csv' (header=true)
+COPY Company FROM 'company.csv' (header=true)
+COPY Merchant FROM 'merchant.csv' (header=true)
+
+// Copy from CSV to edge tables
+COPY TransactedWith FROM 'transacted_with.csv' 
+COPY BelongsTo FROM 'belongs_to.csv'
+COPY LocatedIn FROM 'located_in.csv'
+```
+
+The queries above require that the empty tables were created beforehand. The `COPY <edge_table> FROM <file>` statement
+writes the data into a Kùzu database. Running the queries on an existing database connection
+results in the graph being saved to a local directory.
+
+## Querying the graph
```

Comment:
Overall, let's maybe update these queries and this is the only part I'd like to review again quickly.

Reply:
The queries have been rewritten. I think the new ones showcase the following much betterL

- Recursive joins with kleene star
- Many-to-many joins
- Patterns and path traversals rather than aggregations